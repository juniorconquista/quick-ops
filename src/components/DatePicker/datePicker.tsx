import React, { memo } from 'react';
import ReactDatePicker from 'react-datepicker';
import { Box, Label, P } from 'quarks-ui';
import styled from 'styled-components';
import { color, variant, VariantArgs } from 'styled-system';

import { DatePickerProps, DatePickerStyledProps } from './datePicker.types';

/* TODO This component was created in the local repository as it has not yet been 
implemented in lib Quarks UI, however in the future it can be changed to be used directly from lib. */

const variants = variant({
    prop: 'styling',
    variants: {
        base: {
            width: '100%',
            '.react-datepicker-wrapper': {
                width: '100%',
                input: {
                    height: '36px',
                    borderRadius: '6px',
                    color: '#6b7480 !important',
                    border: '1px solid',
                    borderColor: '#dfe2e6',
                    p: '12px 6px',
                    display: 'flex',
                    alignItems: 'center',
                    bg: 'white.default',
                    cursor: 'pointer',
                    width: '100%',
                    transition: 'boxShadow .25s ease',
                    fontFamily: 'Nunito Sans, sans-serif',
                    '&:hover, &:focus': {
                        boxShadow: '0 0 2px 3px #6b748090',
                        outline: 'none',
                    },
                },
            },
            '.react-datepicker__current-month, .react-datepicker__day-name, .react-datepicker__day': {
                fontFamily: 'Nunito Sans, sans-serif',
                color: 'text.default',
            },
            '.react-datepicker__triangle': {
                display: 'none',
            },
            '.react-datepicker__header': {
                bg: 'white.default',
            },
            '.react-datepicker__month': {
                bg: 'white.contrast',
                margin: '0px',
                p: '0.4rem',
                overflow: 'hidden',
            },
            '.react-datepicker__day': {
                '&:hover': {
                    bg: 'white.default',
                    color: 'text.default',
                },
            },
            '.react-datepicker__day--selected': {
                bg: 'primary.default',
                color: '#FFFF',
                '&:hover': {
                    bg: 'primary.default',
                    color: '#FFFF',
                },
            },
        },
        get success() {
            const base: { [key: string]: any } = this.base;
            return {
                ...base,
                '.react-datepicker-wrapper': {
                    width: '100%',
                    input: {
                        ...base['.react-datepicker-wrapper'].input,
                        borderColor: 'success.default',
                        color: 'success.default',
                        '&:hover, &:focus': {
                            boxShadow: (props: {
                                colors: { success: { default: string } };
                            }) =>
                                `0 0 2px 3px ${props.colors.success.default}90`,
                            outline: 'none',
                        },
                    },
                },
            };
        },
        get danger() {
            const base: { [key: string]: any } = this.base;
            return {
                ...base,
                '.react-datepicker-wrapper': {
                    width: '100%',
                    input: {
                        ...base['.react-datepicker-wrapper'].input,
                        borderColor: 'danger.default',
                        color: 'danger.default',
                        '&:hover, &:focus': {
                            boxShadow: (props: {
                                colors: { danger: { default: string } };
                            }) =>
                                `0 0 2px 3px ${props.colors.danger.default}90`,
                            outline: 'none',
                        },
                    },
                },
            };
        },
    },
} as VariantArgs);

const Style: React.FC<DatePickerStyledProps> = styled.div<
    DatePickerStyledProps
>`
    ${variants}
`;

const DatePicker: React.FC<DatePickerProps> = ({
    styling = 'base',
    label,
    error,
    name,
    ...props
}) => (
    <Box styling="column" width="100%" alignItems="flex-start">
        {label && (
            <Label
                m="12px 0px"
                fontWeight="bold"
                color={error ? 'danger.default' : 'text.contrast'}
            >
                {label}
            </Label>
        )}
        <Style styling={styling}>
            <ReactDatePicker id={name} {...props} />
        </Style>
        {error && (
            <P styling="base" mt="12px" color="danger.default">
                {error}
            </P>
        )}
    </Box>
);

export default memo(DatePicker);
