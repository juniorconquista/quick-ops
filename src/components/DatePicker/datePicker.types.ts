import { ReactDatePickerProps } from 'react-datepicker';

export type DatePickerStyledProps = {
    /** Each styling variant has unique properties that are inherited through the type informed */
    styling?: 'base' | 'success' | 'danger';
};

export interface DatePickerProps extends ReactDatePickerProps {
    styling?: 'base' | 'success' | 'danger';
    label?: string;
    error?: string | boolean;
}
