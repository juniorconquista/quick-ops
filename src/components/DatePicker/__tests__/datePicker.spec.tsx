import React from 'react';
import { cleanup, render } from '@testing-library/react';

import ThemeProvider from '../../../utils/ThemeProviderTest';

import 'jest-styled-components';
import '@testing-library/jest-dom/extend-expect';

import DatePicker from '../datePicker';

const props = {
    onChange: jest.fn(),
};

describe('DatePicker component', () => {
    afterEach(() => {
        jest.clearAllMocks();
        cleanup();
    });
    test('basic snapshot', () => {
        const container = render(
            <ThemeProvider>
                <DatePicker styling="base" {...props} />
            </ThemeProvider>,
        );
        expect(container.asFragment()).toMatchSnapshot();
    });

    it('should inherit the styling properties of the success variant', () => {
        render(
            <ThemeProvider>
                <DatePicker styling="success" {...props} />
            </ThemeProvider>,
        );
    });

    it('should inherit the styling properties of the danger variant', () => {
        render(
            <ThemeProvider>
                <DatePicker styling="danger" {...props} />
            </ThemeProvider>,
        );
    });
});
