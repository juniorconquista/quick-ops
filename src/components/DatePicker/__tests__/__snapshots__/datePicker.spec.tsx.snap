// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePicker component basic snapshot 1`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  width: 100%;
}

.c0:focus {
  outline: none;
}

.c1 {
  width: 100%;
}

.c1 .react-datepicker-wrapper {
  width: 100%;
}

.c1 .react-datepicker-wrapper input {
  height: 36px;
  border-radius: 6px;
  color: #6b7480 !important;
  border: 1px solid;
  border-color: #dfe2e6;
  padding: 12px 6px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #25262a;
  cursor: pointer;
  width: 100%;
  -webkit-transition: boxShadow .25s ease;
  transition: boxShadow .25s ease;
  font-family: Nunito Sans,sans-serif;
}

.c1 .react-datepicker-wrapper input:hover,
.c1 .react-datepicker-wrapper input:focus {
  box-shadow: 0 0 2px 3px #6b748090;
  outline: none;
}

.c1 .react-datepicker__current-month,
.c1 .react-datepicker__day-name,
.c1 .react-datepicker__day {
  font-family: Nunito Sans,sans-serif;
  color: #f8f9fa;
}

.c1 .react-datepicker__triangle {
  display: none;
}

.c1 .react-datepicker__header {
  background-color: #25262a;
}

.c1 .react-datepicker__month {
  background-color: #313438;
  margin: 0px;
  padding: 0.4rem;
  overflow: hidden;
}

.c1 .react-datepicker__day:hover {
  background-color: #25262a;
  color: #f8f9fa;
}

.c1 .react-datepicker__day--selected {
  background-color: #51c1c3;
  color: #FFFF;
}

.c1 .react-datepicker__day--selected:hover {
  background-color: #51c1c3;
  color: #FFFF;
}

<div
    class="c0"
    width="100%"
  >
    <div
      class="c1"
    >
      <div
        class="react-datepicker-wrapper"
      >
        <div
          class="react-datepicker__input-container"
        >
          <input
            class=""
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
